### PR TITLE
Split HTTP1 and HTTP2 parser tests

### PR DESCRIPTION
**Pull request by alegz-filomaphitskay**
Contributes to #1297

What has been done:
- h1/h2 parser tests were splitted in different files;
- added little framework for building h2 tests requests;
- added many new tests for h2 that duplicated h1 tests;
- some h2 parser implementation fixes.

### CODE REVIEW

File: fw/http_parser.c [hunk 1]
```diff
@@ -19,7 +19,6 @@
  * Temple Place - Suite 330, Boston, MA 02111-1307, USA.
  */
 #include <linux/ctype.h>
-#include <linux/kernel.h>
```
Comment (alegz-filomaphitskay) [inline fw/http_parser.c:22]: Removed during fixing bug with disappearing debug messages

File: fw/t/unit/test_http_parser_common.h [hunk 2]
```diff
@@ -0,0 +1,699 @@
...
+	BLOCK_MACRO(head "'dummy'");				\
+	BLOCK_MACRO(head "W/ \"dummy\"");			\
+	BLOCK_MACRO(head "w/\"dummy\"");			\
+	BLOCK_MACRO(head "\"\x00\"");				\
+	BLOCK_MACRO(head "\"\x0F\"");				\
+	BLOCK_MACRO(head "\"\x7F\"");				\
+	BLOCK_MACRO(head "\" \"");				\
+	BLOCK_MACRO(head "\"\"\"")
+
+
```
Comment (krizhanovsky) [inline fw/t/unit/test_http_parser_common.h:108]: There are many places of multiple empty lines - we use not more that 1 empty line in C/C++ code, so could you please adjust empty lines.

**Review by krizhanovsky — CHANGES_REQUESTED** [review]
Quite a big work and many h2 parser bugs are fixes - well done! I didn't find any serious issues. The main issue with the PR is that the most crucial fixes in the parser were done in single commit with almost not comments, so there are many questions (or maybe issues), which must be discussed and resolved.

**Review by alegz-filomaphitskay — COMMENTED** [review]
Fixed cosmetic issues.

File: fw/http_parser.c [hunk 3]
```diff
@@ -6178,7 +6192,7 @@ __h2_req_parse_accept(TfwHttpReq *req, unsigned char *data, size_t len,
 		});
 		c = *(p + __fsm_sz);
 		if (c == '/')
-			__FSM_H2_I_MOVE_n(Req_I_Slash, __fsm_sz + 1);
+			__FSM_H2_I_MOVE_NEQ(Req_I_Slash, __fsm_sz + 1);
```
[the remark this answers is not in the document]
Comment (alegz-filomaphitskay) [reply fw/http_parser.c:6195]: Done. Added for Req_I_AfterText and Req_I_AfterStar.

File: fw/http_parser.c [hunk 4]
```diff
@@ -7641,14 +7679,15 @@ __h2_req_parse_m_override(TfwHttpReq *req, unsigned char *data, size_t len,
 	}
 
 	__FSM_STATE(I_Meth_Unknown) {
-		__FSM_I_MATCH_MOVE(token, I_Meth_Unknown);
-		req->method_override = _TFW_HTTP_METH_UNKNOWN;
+		__FSM_I_MATCH_MOVE_finish(token, I_Meth_Unknown, {
+			req->method_override = _TFW_HTTP_METH_UNKNOWN;
+		});
```
[the remark this answers is not in the document]
Comment (alegz-filomaphitskay) [reply fw/http_parser.c:7684]: It's my mistake. I changed that place.

File: fw/http_parser.c [hunk 5]
```diff
@@ -7641,14 +7679,15 @@ __h2_req_parse_m_override(TfwHttpReq *req, unsigned char *data, size_t len,
...
 		__FSM_H2_I_MOVE_n(I_EoT, __fsm_sz);
 	}
 
 	__FSM_STATE(I_EoT) {
 		if (IS_TOKEN(c))
-			__FSM_H2_I_MOVE(I_Meth_Unknown);
+			__FSM_H2_I_MOVE_n(I_Meth_Unknown, 0);
```
[the remark this answers is not in the document]
Comment (alegz-filomaphitskay) [reply fw/http_parser.c:7690]: Done.

File: fw/http_parser.c [hunk 6]
```diff
@@ -6252,7 +6268,9 @@ __h2_req_parse_accept(TfwHttpReq *req, unsigned char *data, size_t len,
 			__FSM_H2_I_MOVE(Req_I_WSAccept);
 		if (c == ';')
 			/* Skip weight parameter. */
-			__FSM_H2_I_MOVE(Req_I_WSAcceptOther);
+			__FSM_H2_I_MOVE_LAMBDA_n(Req_I_WSAcceptOther, 1, {
+				__FSM_EXIT(CSTR_NEQ);
+			});
```
[the remark this answers is not in the document]
Comment (alegz-filomaphitskay) [reply fw/http_parser.c]: Yes. You are right! Thanks!

File: fw/http_parser.c [hunk 7]
```diff
@@ -5926,68 +5917,78 @@ do {									\
 		__FSM_EXIT(CSTR_EQ);					\
 	}, curr_st, next_st)
 
-#define H2_TRY_STR_LAMBDA_BY_REF_finish(str, lambda1, lambda2, finish, state)\
+/**
+ * The very similar to @H2_TRY_STR_2LAMBDA(), but uses referrers for states.
+ */
+#define H2_TRY_STR_BY_REF(str, curr_st, next_st)			\
```
Comment (alegz-filomaphitskay) [inline fw/http_parser.c:5923]: Perhaps, only one macro definition should be left - H2_TRY_STR_2LAMBDA or H2_TRY_STR_BY_REF?

File: fw/http_sess.c [hunk 8]
```diff
@@ -113,6 +113,16 @@ typedef struct {
...
+void tfw_http_sess_redir_mark_enable(void)
+{
+	redir_mark_enabled = true;
+}
+
+void tfw_http_sess_redir_mark_disable(void)
+{
+	redir_mark_enabled = false;
+}
+
```
[the remark this answers is not in the document]
Comment (alegz-filomaphitskay) [reply fw/http_sess.c]: Ok. I removed that changes.

File: fw/t/unit/test_http_parser_common.h [hunk 9]
```diff
@@ -0,0 +1,710 @@
...
+    TfwFrameType subtype;	// used only for FUZZ_REQ_H2 cases
+} TfwFrameRec;
+
+#define ALLOWED_FRAMES_CNT 2
+
+static TfwFrameRec frames[ALLOWED_FRAMES_CNT];
+static unsigned int frames_cnt = 0;
+static unsigned int frames_max_sz = 0;
+static unsigned int frames_total_sz = 0;
+static TfwHeaderRec tmp_header_rec __attribute__((unused));
```
Comment (alegz-filomaphitskay) [inline fw/t/unit/test_http_parser_common.h:138]: I will remove that struct in [#1612](https://github.com/tempesta-tech/tempesta/pull/1612)

File: fw/http_parser.c [hunk 10]
```diff
@@ -5957,36 +5962,45 @@ do {									\
...
-		if (likely(fin))					\
-			__FSM_EXIT(CSTR_NEQ);				\
-		parser->_i_st = &&curr_st;				\
-		__msg_field_fixup_pos(fld, p, __fsm_n);			\
-		__FSM_I_field_chunk_flags(fld, TFW_STR_HDR_VALUE);	\
-		__FSM_EXIT(CSTR_POSTPONE);				\
-	}
-
-#define H2_TRY_STR_LAMBDA_fixup(str, fld, lambda, curr_st, next_st)	\
+#define H2_TRY_STR_3LAMBDA_fixup(str, fld, lambda1, lambda2, lambda3,           \
```
[the remark this answers is not in the document]
Comment (krizhanovsky) [reply fw/http_parser.c]: Something like `full_match_lamda`, `full_match_fin_lambda`, `some_match_fin_lambda`

File: fw/http_parser.c (hunk 7 above)
Comment (krizhanovsky) [reply fw/http_parser.c:5923]: Yeah, good point: `H2_TRY_STR_BY_REF` can be simply expressed via `H2_TRY_STR_2LAMBDA`

File: fw/http_parser.c [hunk 11]
```diff
@@ -5957,36 +5962,45 @@ do {									\
...
+					p += __fsm_n;				\
+					goto next_st;				\
+				}						\
+				if (likely(fin))				\
+					lambda2;				\
+				parser->_i_st = &&next_st;			\
+				__FSM_EXIT(CSTR_POSTPONE);			\
+			}							\
+			__msg_field_fixup_pos(fld, p, __fsm_n);			\
+			__FSM_I_field_chunk_flags(fld, TFW_STR_HDR_VALUE);	\
```
[the remark this answers is not in the document]
Comment (krizhanovsky) [reply fw/http_parser.c:5966]: Yes, you're right: we have such lambda1 in `__h2_req_parse_content_type()` state `I_ContTypeParam`. Could you please add a comment for `lambda1` in `H2_TRY_STR_3LAMBDA_fixup` that it might return with an error code and we don't need to fixup the current data chunk.

File: fw/http_parser.c [hunk 12]
```diff
@@ -5926,68 +5917,78 @@ do {									\
...
+		if (likely(chunk->len == (str)->len)) {				\
+			lambda1;						\
+			TRY_STR_INIT();						\
+			__msg_field_fixup_pos(fld, p, __fsm_n);			\
+			__FSM_I_field_chunk_flags(fld, TFW_STR_HDR_VALUE);	\
+			if (__data_off(p + __fsm_n) < len) {			\
+				p += __fsm_n;					\
+				goto next_st;					\
+			}							\
+			parser->_i_st = &&next_st;				\
```
Comment (krizhanovsky) [inline fw/http_parser.c:5971]: Why did you move the line before `if (fin)`? Probably we need a comment here to note the importance of the code placement.

File: fw/http_parser.c [hunk 13]
```diff
@@ -6565,10 +6579,14 @@ __h2_req_parse_content_type(TfwHttpMsg *hm, unsigned char *data, size_t len,
...
 			TFW_STR_STRING("multipart/form-data");
-		H2_TRY_STR_LAMBDA_fixup(&s_multipart_form_data, &parser->hdr, {
-			__set_bit(TFW_HTTP_B_CT_MULTIPART, req->flags);
-			__FSM_EXIT(CSTR_EQ);
-		}, I_ContTypeMediaType, I_ContTypeMaybeMultipart);
+		H2_TRY_STR_3LAMBDA_fixup(&s_multipart_form_data, &parser->hdr, {}, {
+				__set_bit(TFW_HTTP_B_CT_MULTIPART, req->flags);
+				__FSM_EXIT(CSTR_EQ);
+			}, {
+				p += __fsm_n;
```
[the remark this answers is not in the document]
Comment (krizhanovsky) [reply fw/http_parser.c]: The comment was about `p += __fsm_n` statement which worked as `lambda3` in `H2_TRY_STR_3LAMBDA_fixup`, i.e. we moved `p` only if we're in the final chunk - I was wondering why don't we move `p` in other cases?

File: fw/http_parser.c [hunk 14]
```diff
@@ -6558,56 +6557,71 @@ __h2_req_parse_content_type(TfwHttpMsg *hm, unsigned char *data, size_t len,
...
+			TFW_STR_STRING("form-data");
+		H2_TRY_STR_2FIN_fixup(&s_form_data, &parser->hdr, {
+				__set_bit(TFW_HTTP_B_CT_MULTIPART, req->flags);
+				__FSM_EXIT(CSTR_EQ);
+			}, {
+				__FSM_EXIT(CSTR_EQ);
+			}, I_ContTypeAfterMultipartSlash, I_ContTypeMaybeMultipart);
+		TRY_STR_INIT();
+		__FSM_H2_I_JMP(I_ContTypeOtherSubtype);
 	}
```
Comment (krizhanovsky) [inline fw/http_parser.c:6587]: Let's not to introduce the new state: frankly the code is still very sophisticated and the change is doubtful from the performance point of view. Let's just clarify the previous change about `p` movement in `H2_TRY_STR_3LAMBDA_fixup`.

File: fw/http_parser.c [hunk 15]
```diff
@@ -6044,7 +6058,7 @@ __h2_req_parse_authority(TfwHttpReq *req, unsigned char *data, size_t len,
 	__FSM_STATE(Req_I_A_v6) {
 		/* See Req_UriAuthorityIPv6 processing. */
 		if (likely(isxdigit(c) || c == ':'))
-			__FSM_H2_I_MOVE_fixup(Req_I_A_v6, 1, TFW_STR_VALUE);
+			__FSM_H2_I_MOVE_NEQ_n_flag(Req_I_A_v6, 1, TFW_STR_VALUE);
```
[the remark this answers is not in the document]
Comment (krizhanovsky) [reply fw/http_parser.c:6048]: Which exactly tests are broken without the fixup-removing change?

At the moment I'm confused with `__msg_hdr_chunk_fixup()` at the end of parsing and it seems for me that the code works for contiguous input data only (note that the current chunking tests aren't ideal since they just splits a contiguous data and chunks are actually not bordered).

**Review by krizhanovsky — CHANGES_REQUESTED** [review]
Looks good, just several questions should be resolved

File: fw/http_parser.c [hunk 16]
```diff
@@ -6149,12 +6150,12 @@ __h2_req_parse_accept(TfwHttpReq *req, unsigned char *data, size_t len,
 			__FSM_EXIT(CSTR_EQ);
 		},  Req_I_AfterTextSlashToken, Req_I_AcceptHtml);
 		TRY_STR_INIT();
-		__FSM_I_JMP(Req_I_Subtype);
+		__FSM_H2_I_JMP(Req_I_Subtype);
```
[the remark this answers is not in the document]
Comment (alegz-filomaphitskay) [reply fw/http_parser.c:6153]: It's my fault. I removed __FSM_H2_I_JMP and instead of it use __FSM_I_JMP.

File: fw/http_parser.c (hunk 11 above)
[the remark this answers is not in the document]
Comment (alegz-filomaphitskay) [reply fw/http_parser.c:5966]: Done

File: fw/http_parser.c (hunk 12 above)
Comment (alegz-filomaphitskay) [reply fw/http_parser.c:5971]: I think that I was wrong here. I considered that there was a sufficient condition to change the state. But for this, the chunk must not be fin.

File: fw/http_parser.c [hunk 17]
```diff
@@ -7276,8 +7273,9 @@ do {									\
 #undef __NEXT_TEMPL_STATE
 
 	__FSM_STATE(I_GMT) {
-		H2_TRY_STR_BY_REF("gmt",
-			&&I_GMT, st[parser->date.type][parser->date.pos + 1]);
+		H2_TRY_STR_LAMBDA("gmt", {
+			__FSM_I_JMP(I_Res);
```
Comment (alegz-filomaphitskay) [inline fw/http_parser.c:7277]: It is known that I_GMT is followed by I_Res.

File: fw/http_parser.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (krizhanovsky) [reply fw/http_parser.c:6549]: Please add this with a comment, what each of the 2 lambdas for `H2_TRY_STR_FULL_OR_PART_MATCH_FIN_fixup()` in `I_ContTypeMediaType` state do, why do we do `p` movement in exactly this way.

File: fw/http_parser.c (hunk 17 above)
Comment (krizhanovsky) [reply fw/http_parser.c:7277]: It seems you're right, but could you please then fix the same place for the HTTP/1 parser and write a 2 `gmt` test cases for both the parsers?

File: fw/http_parser.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (krizhanovsky) [reply fw/http_parser.c:6023]: So what is wrong with the scenario `authority: [::1]:5001` from #1566? You do the opposite change, so we need a very precise understanding on what was wrong with that change, i.e. were the conclusions about the test case processing were wrong.

**Review by krizhanovsky — COMMENTED** [review]
The only major thins is to analyze what was wrong with #1566 analytics and the tests reporting

File: fw/http_parser.c (hunk 13 above)
[the remark this answers is not in the document]
Comment (alegz-filomaphitskay) [reply fw/http_parser.c:6549]: Done.

File: fw/http_parser.c [hunk 18]
```diff
@@ -3141,8 +3150,13 @@ do {									\
...
 
 	__FSM_STATE(I_GMT) {
+
 		TRY_STR_BY_REF("gmt", &&I_GMT,
-		               st[parser->date.type][parser->date.pos + 1]);
+			/*
+			 * The st[][]-table is not used because it is known
+			 * that I_GMT is followed by I_Res.
+			 */
+			&&I_Res);
```
Comment (alegz-filomaphitskay) [inline fw/http_parser.c:3159]: May be better to revert changes to st[][]-table usage?

File: fw/http_parser.c [hunk 19]
```diff
@@ -7276,8 +7314,13 @@ do {									\
 #undef __NEXT_TEMPL_STATE
 
 	__FSM_STATE(I_GMT) {
-		H2_TRY_STR_BY_REF("gmt",
-			&&I_GMT, st[parser->date.type][parser->date.pos + 1]);
+		H2_TRY_STR_LAMBDA("gmt", {
```
Comment (alegz-filomaphitskay) [inline fw/http_parser.c:7317]: May be better to revert changes to st[][]-table usage?

File: fw/t/unit/test_http2_parser.c [hunk 20]
```diff
@@ -0,0 +1,2684 @@
...
+#define EXPECT_FOR_REQ_H2_METHOD_OVERRIDE(METHOD)				\
+	FOR_REQ_H2(								\
+	    HEADERS_FRAME_BEGIN();						\
+		HEADER(STR(":method"), STR("POST"));				\
+		HEADER(STR(":scheme"), STR("https"));				\
+		HEADER(STR(":path"), STR("/"));					\
+		HEADER(STR(":authority"), STR("example.com"));			\
+		HEADER(STR("x-method-override"), STR(#METHOD));			\
+	    HEADERS_FRAME_END();						\
+	);									\
```
Comment (alegz-filomaphitskay) [inline fw/t/unit/test_http2_parser.c:2334]: Must be fixed like `);` --> `)`

File: fw/t/unit/test_http2_parser.c [hunk 21]
```diff
@@ -0,0 +1,2684 @@
...
+{
+#define EXPECT_BLOCK_BODYLESS_REQ_H2(METHOD)					\
+	EXPECT_BLOCK_REQ_H2(							\
+	    HEADERS_FRAME_BEGIN();						\
+		HEADER(STR(":method"), STR(#METHOD));				\
+		HEADER(STR(":scheme"), STR("https"));				\
+		HEADER(STR(":path"), STR("/filename"));				\
+		HEADER(STR("content-length"), STR("0"));			\
+	    HEADERS_FRAME_END();						\
+	);									\
```
Comment (alegz-filomaphitskay) [inline fw/t/unit/test_http2_parser.c:903]: Must be fixed like `);` --> `)`

File: fw/t/unit/test_http2_parser.c (hunk 20 above)
Comment (krizhanovsky) [reply fw/t/unit/test_http2_parser.c:2334]: Yeah, please fix both the cases

File: fw/http_parser.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (krizhanovsky) [reply fw/http_parser.c:6040]: OK, looks good. My understanding of https://github.com/tempesta-tech/tempesta/pull/1566/commits/0d98a42425f2b074a21bd50deea25926aa83de34 is that to prevent `[[::1]:5001` parsed from 2 chunks `[:`, `:1]:5001` we shouldn't fixup `[` in `Req_I_A_Start`.

In general, explicit fixups were introduced to explicitly separate contiguous strings, e.g. `Host: foo.com`, to separate parts for easier operation, in this case there header name and the header value. In normal case we fixup a chunk only when there is not enough data (returning TFW_POSTPONE). We should not create extra chunks because this hurt performance. So the statement that we either use fixup or non-fixup functions is wrong.

**Review by krizhanovsky — APPROVED** [review]
Looks good and thank you for the detailed explanation of the issue with the fixup logic.

During the PR tests I found a bug in HTTP/2 redirection mark parsing and improved code comments about the fixup logic. Please review https://github.com/tempesta-tech/tempesta/pull/1623 and merge it on your own once this #1597 is merged.

File: fw/http_parser.c (hunk 15 above)
[the remark this answers is not in the document]
Comment (alegz-filomaphitskay) [reply fw/http_parser.c:6040]: > We should not create extra chunks because this hurt performance. So the statement that we either use fixup or non-fixup functions is wrong.

Yes, I totally agreed with this a couple of posts ago:

> This version of code (before my changes) can be fixed with such changes in the last step of parsing:
> ```
> 		if (likely(c == ']')) {
> 			__msg_hdr_chunk_fixup(p, 1);
> 			__msg_chunk_flags(TFW_STR_HDR_VALUE | TFW_STR_VALUE);
> ```
> But this version of the code with _fixup-macros has one disadvantage: we always fixup data by one chunk at a time, no matter what the actual size of the chunks is.
